src/remote_tables/pushdown_visitor.rs

```diff
@@ -0,0 +1,204 @@
+use datafusion_expr::{BinaryExpr, Expr, Operator};
+
+pub struct FilterPushdown<T: FilterPushdownVisitor> {
+    pub source: T,
+    pub pushdown_supported: bool,
+    // LIFO stack for keeping the intermediate SQL expression results to be used in interpolation
+    // of the parent nodes. After a successful visit, it should contain exactly one element, which
+    // represents the complete SQL statement corresponding to the given expression.
+    pub sql_exprs: Vec<String>,
+}
+
+pub struct PostgresFilterPushdown {}
+pub struct SQLiteFilterPushdown {}
+pub struct MySQLFilterPushdown {}
+
+impl FilterPushdownVisitor for PostgresFilterPushdown {}
+
+impl FilterPushdownVisitor for SQLiteFilterPushdown {
+    fn op_to_sql(&self, op: Operator) -> Option<String> {
+        match op {
+            Operator::RegexMatch
+            | Operator::RegexIMatch
+            | Operator::RegexNotMatch
+            | Operator::RegexNotIMatch
+            | Operator::BitwiseXor => None,
+            _ => Some(op.to_string()),
+        }
+    }
+}
+
+impl FilterPushdownVisitor for MySQLFilterPushdown {
+    fn op_to_sql(&self, op: Operator) -> Option<String> {
+        match op {
+            // TODO: see if there's a way to convert the non-case sensitive match
+            Operator::RegexIMatch | Operator::RegexNotIMatch => None,
+            Operator::RegexMatch => Some("RLIKE".to_string()),
+            Operator::RegexNotMatch => Some("NOT RLIKE".to_string()),
+            Operator::BitwiseXor => Some("^".to_string()),
+            _ => Some(op.to_string()),
+        }
+    }
+}
+
+pub trait FilterPushdownVisitor {
+    fn scalar_value_to_sql(&self, value: &ScalarValue) -> Option<String> {
+        match value {
+            ScalarValue::Utf8(Some(val)) => Some(format!("'{}'", val)),
+            _ => Some(format!("{}", value)),
+        }
+    }
+
+    fn op_to_sql(&self, op: Operator) -> Option<String> {
+        Some(op.to_string())
+    }
+}
+
+impl<T: FilterPushdownVisitor> ExpressionVisitor for FilterPushdown<T> {
+    fn pre_visit(mut self, expr: &Expr) -> Result<Recursion<Self>> {
+        match expr {
+            Expr::Column(_) | Expr::Literal(_) => {}
+            Expr::BinaryExpr(BinaryExpr { op, .. }) => {
+                // Check if operator pushdown supported; left and right expressions will be checked
+                // through further recursion.
+                if self.source.op_to_sql(*op).is_none() {
+                    return Ok(Recursion::Stop(self));
+                }
+            }
+            _ => {
+                // Expression is not supported, no need to visit any remaining nodes
+                self.pushdown_supported = false;
+                return Ok(Recursion::Stop(self));
+            }
+        };
+        Ok(Recursion::Continue(self))
+    }
+
+    fn post_visit(mut self, expr: &Expr) -> Result<Self> {
+        match expr {
+            Expr::Column(col) => self.sql_exprs.push(col.name.clone()),
```

Comment:
We should consider quoting column names here (in PG, the identifiers are quoted with double quotes, e.g. `'"' + col.replace('"', '""') + '"'"`); in MySQL they use backticks (`)

Reply:
Source-compatible identifier quotation and escaping added now.